aggregator/src/aggregator/aggregation_job_writer.rs

```diff
@@ -699,92 +515,81 @@ where
             )
             .collect();
 
-        // Read all relevant batches and report counts from the datastore.
-        let (batches, batches_with_unaggregated_reports) = try_join!(
-            try_join_all(
-                updates
-                    .by_batch_identifier_index
-                    .keys()
-                    .map(|batch_identifier| {
-                        tx.get_batch::<SEED_SIZE, Q, A>(
-                            updates.task.id(),
-                            batch_identifier,
-                            aggregation_parameter,
-                        )
-                    })
-            ),
-            try_join_all(updates.by_batch_identifier_index.keys().map(
-                |batch_identifier| async move {
-                    if let Some(batch_interval) = Q::to_batch_interval(batch_identifier) {
-                        if tx
-                            .interval_has_unaggregated_reports(updates.task.id(), batch_interval)
-                            .await?
-                        {
-                            return Ok::<_, Error>(Some(batch_identifier.clone()));
-                        }
-                    }
-                    Ok(None)
-                },
-            )),
-        )?;
-
-        let batches: HashMap<_, _> = batches
-            .into_iter()
-            .flat_map(|batch: Option<Batch<SEED_SIZE, Q, A>>| {
-                batch.map(|b| (b.batch_identifier().clone(), (Operation::Update, b)))
-            })
-            .collect();
+        // Read all relevant batch aggregations from the datastore.
+        let batch_aggregation_ord = thread_rng().gen_range(0..writer.batch_aggregation_shard_count);
```

Comment:
Dumb question: what if the relevant batch aggregation was previously written with a different `batch_aggregation_ord` than what is generated here?

More broadly, it'd help to have a high level explanation of how batch aggregation sharding works. What's the data model? For a given batch, can we have multiple batch aggregations in each shard, or just one?

Reply:
Here's what I'm putting together. Let's suppose that the aggregation job creator has found some unaggregated reports and decided to create a new aggregation job.

- `AggregationJobWriter::write` will choose a random shard to consult each time it's called
- If it finds a relevant batch aggregation in the chosen shard, then either that shard gets updated with the new aggregation job and report aggregations, or `AggregationJobWriter` will refuse to do that for a shard that's in the `Collected` or `Scrubbed` states.
- If it does not find a relevant batch aggregation in the chosen shard, `AggregationJobWriter` creates a new BA in state `Aggregating`.
- As a result, for any given `BatchIdentifier`, we can have at most one `BatchAggregation` per shard. 

But suppose this happens:

- `batch_aggregation_shard_count` is 2, and we are running a time interval task.
- A bunch of reports arrive, timestamped between `[t0, t1)`.
- The aggregation job creator runs, and `AggregationJobWriter::write` winds up creating 1 batch aggregation in shard 0.
- The aggregation jobs run to completion, and the BA in shard 0 is updated into the `Collected` state.
- More reports arrive, also timestamped between `[t0, t1)`.
- The aggregation job creator runs again, finds those unaggregated reports, and creates a new aggregation job.
- This time, `AggregationJobWriter::write` chooses shard 1, and finds no relevant existing batch aggregation.

At this point, I'm pretty sure `WriteState::update_batch_aggregations` will create a brand new `BatchAggregation` in state `Aggregating`, and then the new reports will be aggregated, when they should get failed with `BatchCollected`.

Does the logic that checks if the batch is already collected need to check whether a relevant batch aggregation exists in _any_ shard, not just the one randomly chosen on this run?

Or maybe the fix here is for `AggregationJobWriter`/`WriteState` to eagerly create a `BatchAggregation` in all shards, even though some of them may never have aggregation jobs or report aggregations assigned to them, so that others parts of the machinery can update _all_ the `BatchAggregations` into `Collected` or `Scrubbed` in lockstep?

n.b. I still have to carefully study the code that drives aggregation jobs and creates/drives collection jobs, so this may well be accounted for elsewhere that I haven't seen yet.

Reply:
> More broadly, it'd help to have a high level explanation of how batch aggregation sharding works. What's the data model? For a given batch, can we have multiple batch aggregations in each shard, or just one?

The data model is: a batch aggregation is identified by a batch (i.e. `(task_id, batch_id, aggregation_param`)) and a shard number (i.e. `ord`). A given batch can have many batch aggregations, but each one has a different `ord`. The `ord` defines the shard number of the particular batch aggregation. So, each shard (or `ord`) can hold exactly one `BatchAggregation`. This was true before this PR, and has not changed with this PR.

> At this point, I'm pretty sure WriteState::update_batch_aggregations will create a brand new BatchAggregation in state Aggregating, and then the new reports will be aggregated, when they should get failed with BatchCollected.

This won't happen: first, the step where the first aggregation job runs to completion will not update the BA to `Collected` (the aggregation job writer does not update batch aggregation states). This will only happen once a collection job is received from the collector, and the collection job is stepped and notices that it can complete the collection job (i.e. there are no outstanding aggregation jobs & no outstanding reports in the batch). At that point, the collection job stepper will update _all_ batch aggregation shards to the `Collected` state, to avoid the problem you describe.

That is,

> Or maybe the fix here is for AggregationJobWriter/WriteState to eagerly create a BatchAggregation in all shards, even though some of them may never have aggregation jobs or report aggregations assigned to them, so that others parts of the machinery can update all the BatchAggregations into Collected or Scrubbed in lockstep?

This is what is already implemented, except that it happens in the collection job driver (for the Leader) or the aggregate share request handler (for the Helper) instead of anywhere an aggregation job is written. See e.g. https://github.com/divviup/janus/blob/79597db649cc01cb18680bfda17b4e95a626aed8/aggregator/src/aggregator/collection_job_driver.rs#L192-L233.

I'm going to write up a description of state handling in the `docs/` directory, as I think this is pretty complicated.

(FWIW, I think we should rename `batch_aggregations` to `batches` as a next step, since `batch_aggregations` stores all of the per-batch information now. But let's not confuse the current discussion!)

Reply:
Those explanations make sense, thank you!